peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+		// actually connected from.  One example of why this can happen
+		// is with NAT.  Only add the address to the address manager if
+		// the addresses agree.
+		if addrmgr.NetAddressKey(&msg.AddrMe) == addrmgr.NetAddressKey(p.na) {
+			p.addrManager.AddAddress(p.na, p.na)
+			p.addrManager.Good(p.na)
+		}
+	}
+}
+
+// PushGetBlocksMsg sends a getblocks message for the provided block locator
+// and stop hash.  It will ignore back-to-back duplicate requests.
+func (p *Peer) PushGetBlocksMsg(locator blockchain.BlockLocator, stopHash *wire.ShaHash) error {
+	// Extract the begin hash from the block locator, if one was specified,
+	// to use for filtering duplicate getblocks requests.
+	// request.
+	var beginHash *wire.ShaHash
+	if len(locator) > 0 {
+		beginHash = locator[0]
+	}
+
+	// Filter duplicate getblocks requests.
+	if p.prevGetBlocksStop != nil && p.prevGetBlocksBegin != nil &&
+		beginHash != nil && stopHash.IsEqual(p.prevGetBlocksStop) &&
+		beginHash.IsEqual(p.prevGetBlocksBegin) {
+
+		log.Tracef("Filtering duplicate [getblocks] with begin "+
+			"hash %v, stop hash %v", beginHash, stopHash)
+		return nil
+	}
+
+	// Construct the getblocks request and queue it to be sent.
+	msg := wire.NewMsgGetBlocks(stopHash)
+	for _, hash := range locator {
+		err := msg.AddBlockLocatorHash(hash)
+		if err != nil {
+			return err
+		}
+	}
+	p.QueueMessage(msg, nil)
+
+	// Update the previous getblocks request information for filtering
+	// duplicates.
+	p.prevGetBlocksBegin = beginHash
+	p.prevGetBlocksStop = stopHash
+	return nil
+}
+
+// PushGetHeadersMsg sends a getblocks message for the provided block locator
+// and stop hash.  It will ignore back-to-back duplicate requests.
+func (p *Peer) PushGetHeadersMsg(locator blockchain.BlockLocator, stopHash *wire.ShaHash) error {
+	// Extract the begin hash from the block locator, if one was specified,
+	// to use for filtering duplicate getheaders requests.
+	var beginHash *wire.ShaHash
+	if len(locator) > 0 {
+		beginHash = locator[0]
+	}
+
+	// Filter duplicate getheaders requests.
+	if p.prevGetHdrsStop != nil && p.prevGetHdrsBegin != nil &&
+		beginHash != nil && stopHash.IsEqual(p.prevGetHdrsStop) &&
+		beginHash.IsEqual(p.prevGetHdrsBegin) {
+
+		log.Tracef("Filtering duplicate [getheaders] with begin "+
+			"hash %v", beginHash)
+		return nil
+	}
+
+	// Construct the getheaders request and queue it to be sent.
+	msg := wire.NewMsgGetHeaders()
+	msg.HashStop = *stopHash
+	for _, hash := range locator {
+		err := msg.AddBlockLocatorHash(hash)
+		if err != nil {
+			return err
+		}
+	}
+	p.QueueMessage(msg, nil)
+
+	// Update the previous getheaders request information for filtering
+	// duplicates.
+	p.prevGetHdrsBegin = beginHash
+	p.prevGetHdrsStop = stopHash
+	return nil
+}
+
+// PushRejectMsg sends a reject message for the provided command, reject code,
+// and reject reason, and hash.  The hash will only be used when the command
```

Comment:
I think it was already there, but might as well fix it while we're here.  There is an extra `and` before `reject reason`.